### DISCUSSION TITLE
v3.0.0a0

### DISCUSSION THREAD

**Discussion by njzjz** [deepmodeling/deepmd-kit#3401] [Announcement]
# DeePMD-kit v3: A multiple-backend framework for deep potentials

We are excited to announce the first alpha version of DeePMD-kit v3. DeePMD-kit v3 allows you to train and run deep potential models on top of TensorFlow or PyTorch. DeePMD-kit v3 also supports the [DPA-2 model](https://arxiv.org/abs/2312.15492), a novel architecture for large atomic models.

### Multiple-backend framework

![image](https://github.com/deepmodeling/deepmd-kit/assets/9496702/6bf132d2-6952-4009-b263-3648641003e4)

DeePMD-kit v3 adds a pluggable multiple-backend framework to provide consistent training and inference experiences between different backends. You can:

- Use the same training data and the input script to train a deep potential model with different backends. Switch backends based on efficiency, functionality, or convenience:
```sh
# Training a model using the TensorFlow backend
dp --tf train input.json
dp --tf freeze

# Training a mode using the PyTorch backend
dp --pt train input.json
dp --pt freeze
```

- Use any model to perform inference via any existing interfaces, including `dp test`, Python/C++/C interface, and third-party packages (dpdata, ASE, LAMMPS, AMBER, Gromacs, i-PI, CP2K, OpenMM, ABACUS, etc). Take an example on LAMMPS:
```sh
# run LAMMPS with a TensorFlow backend model
pair_style deepmd frozen_model.pb
# run LAMMPS with a PyTorch backend model
pair_style deepmd frozen_model.pth
# Calculate model deviation using both models
pair_style deepmd frozen_model.pb frozen_model.pth out_file md.out out_freq 100
```

- Convert models between backends, using `dp convert-backend`, if both backends support a model:
```sh
dp convert-backend frozen_model.pb frozen_model.pth
dp convert-backend frozen_model.pth frozen_model.pb
```

- Add a new backend to DeePMD-kit much more quickly if you want to contribute to DeePMD-kit.

### PyTorch backend: a backend designed for large atomic models and new research

We added the PyTorch backend in DeePMD-kit v3 to support the development of new models, especially for large atomic models.

#### DPA-2 model: Towards a universal large atomic model for molecular and material simulation

[DPA-2 model](https://arxiv.org/abs/2312.15492) is a novel architecture for [Large Atomic Model](https://github.com/deepmodeling/community/discussions/32) (LAM) and can accurately represent a diverse range of chemical systems and materials, enabling high-quality simulations and predictions with significantly reduced efforts compared to traditional methods. The DPA-2 model is only implemented in the PyTorch backend. An example configuration is in the `examples/water/dpa2` directory.

The DPA-2 descriptor includes two primary components: `repinit` and `repformer`. The detailed architecture is shown in the following figure.

![DPA-2](https://github.com/deepmodeling/deepmd-kit/assets/9496702/9f342b7d-5b68-4dcf-9df2-0fbadb58cec3)

#### Training strategies for large atomic models

The PyTorch backend has supported multiple training strategies to develop large atomic models.

**Parallel training**: Large atomic models have a number of hyper-parameters and complex architecture, so training a model on multiple GPUs is necessary. Benefiting from the PyTorch community ecosystem, the parallel training for the PyTorch backend can be driven by [`torchrun`](https://pytorch.org/docs/stable/elastic/run.html), a launcher for distributed data parallel.

```sh
torchrun --nproc_per_node=4 --no-python dp --pt train input.json
```

**Multi-task training**: Large atomic models are trained against data in a wide scope and at different DFT levels, which requires multi-task training. The PyTorch backend supports multi-task training, sharing the descriptor between different  An example is given in `examples/water_multi_task/pytorch_example/input_torch.json`.

**Finetune**: Fine-tune is useful to train a pre-train large model on a smaller, task-specific dataset. The PyTorch backend has supported `--finetune` argument in the `dp --pt train` command line.

#### Developing new models using Python and dynamic graphs

Researchers may feel pain about the static graph and the custom C++ OPs from the TensorFlow backend, which sacrifices research convenience for computational performance. The PyTorch backend has a well-designed code structure written using the dynamic graph, which is currently 100% written with the Python language, making extending and debugging new deep potential models easier than the static graph.

#### Supporting traditional deep potential models

People may still want to use the traditional models already supported by the TensorFlow backend in the PyTorch backend and compare the same model among different backends. We almost rewrote all of the traditional models in the PyTorch backend, which are listed below:

- Features supported:
  - Descriptor: `se_e2_a`, `se_e2_r`, `se_atten`, `hybrid`;
  - Fitting: energy, dipole, polar, fparam/apram support
  - Model: `standard`, DPRc
  - Python inference interface
  - C++ inference interface for energy only
  - TensorBoard
- Features not supported yet:
  - Descriptor: `se_e3`, `se_atten_v2`, `se_e2_a_mask`
  - Fitting: `dos`
  - Model: `linear_ener`, DPLR, `pairtab`, `linear_ener`, `frozen`, `pairwise_dprc`, ZBL, Spin
  - Model compression
  - Python inference interface for DPLR
  - C++ inference interface for tensors and DPLR
  - Paralleling training using Horovod
- Features not planned:
  - Descriptor: `loc_frame`, `se_e2_a` + type embedding, `se_a_ebd_v2`
  - NVNMD

> [!WARNING] 
> As part of an alpha release, the PyTorch backend's API or user input arguments may change before the first stable version.

### DP backend and format: reference backend for other backends

DP is a reference backend for development that uses pure NumPy to implement models without using any heavy deep-learning frameworks. It cannot be used for training but only for Python inference. As a reference backend, it is not aimed at the best performance but only the correct results. The DP backend uses HDF5 to store model serialization data, which is backend-independent.
The DP backend and the serialization data are used in the unit test to ensure different backends have consistent results and can be converted between each other.
In the current version, the DP backend has a similar supporting status to the PyTorch backend, while DPA-1 and DPA-2 are not supported yet.

### Authors

The above highlights were mainly contributed by
- Hangrui Bi (@20171130), in #3180
- Chun Cai (@caic99), in #3180
- Junhan Chang (@TablewareBox), in #3180
- Yiming Du (@nahso), in #3180
- Guolin Ke (@guolinke), in #3180
- Xinzijian Liu (@zjgemi), in #3180
- Anyang Peng (@anyangml), in #3362, #3192, #3212, #3210, #3248, #3266, #3281, #3296, #3309, #3314, #3321, #3327, #3338, #3351, #3376, #3385
- Xuejian Qin (@qin2xue3jian4), in #3180
- Han Wang (@wanghan-iapcm), in #3188, #3190, #3208, #3184, #3199, #3202, #3219, #3225, #3232, #3235, #3234, #3241, #3240, #3246, #3260, #3274, #3268, #3279, #3280, #3282, #3295, #3289, #3340, #3352, #3357, #3389, #3391, #3400
- Jinzhe Zeng (@njzjz), in #3171, #3173, #3174, #3179, #3193, #3200, #3204, #3205, #3333, #3360, #3364, #3365, #3169, #3164, #3175, #3176, #3187, #3186, #3191, #3195, #3194, #3196, #3198, #3201, #3207, #3226, #3222, #3220, #3229, #3226, #3239, #3228, #3244, #3243, #3213, #3249, #3250, #3254, #3247, #3253, #3271, #3263, #3258, #3276, #3285, #3286, #3292, #3294, #3293, #3303, #3304, #3308, #3307, #3306, #3316, #3315, #3318, #3323, #3325, #3332, #3331, #3330, #3339, #3335, #3346, #3349, #3350, #3310, #3356, #3361, #3342, #3348, #3358, #3366, #3374, #3370, #3373, #3377, #3382, #3383, #3384, #3386, #3390, #3395, #3394, #3396, #3397
- Chengqian Zhang (@Chengqian-Zhang), in #3180
- Duo Zhang (@iProzd), in #3180, #3203, #3245, #3261, #3262, #3355, #3367, #3359, #3371, #3387, #3388, #3380, #3378
- Xiangyu Zhang (@CaRoLZhangxy), in #3162, #3287, #3337, #3375, #3379

## Breaking changes

- Python 3.7 support is dropped. by @njzjz in #3185
- We require all model files to have the correct filename extension for all interfaces so a corresponding backend can load them. TensorFlow model files must end with `.pb` extension.
- Python class `DeepTensor` (including `DeepDiople` and `DeepPolar`) now returns atomic tensor in the dimension of `natoms` instead of `nsel_atoms`. by @njzjz in #3390
- For developers: the Python module structure is fully refactored. The old `deepmd` module was moved to `deepmd.tf` without other API changes, and `deepmd_utils` was moved to `deepmd` without other API changes. by @njzjz in #3177, #3178

### Enhancement

* Neighbor stat for the TensorFlow backend is 80x accelerated. by @njzjz in #3275
* i-PI: remove normalize_coord by @njzjz in #3257
* LAMMPS: fix_dplr.cpp delete redundant setup and set atom->image when pre_force by @shiruosong in #3344, #3345
* Bump scikit-build-core to 0.8 by @njzjz in #3369
* Bump LAMMPS to stable_2Aug2023_update3 by @njzjz in #3399
* Add fparam/aparam support for fine-tune by @njzjz in #3313
* TF: remove freeze warning for optional nodes by @njzjz in #3381

### CI/CD
* Build macos-arm64 wheel on M1 runners by @njzjz in #3206
* Other improvements and fixes to GitHub Actions by @njzjz in #3238, #3283, #3284, #3288, #3290, #3326
* Enable docstring code format by @njzjz in #3267

### Bugfix
* Fix TF 2.16 compatibility by @njzjz in #3343
* Detect version in advance before building deepmd-kit-cu11 by @njzjz in #3172
* C API: change the required shape of electric field to nloc * 3 by @njzjz in #3237

## New Contributors
* @anyangml made their first contribution in #3192
* @shiruosong made their first contribution in #3344

**Full Changelog**: https://github.com/deepmodeling/deepmd-kit/compare/v2.2.8...v3.0.0a0

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmodeling/deepmd-kit/releases/tag/v3.0.0a0'>v3.0.0a0</a>.</em>

**Comment by ShangChien** [comment]
Great works 🎉
Question about section of DP backend: How to use pure Numpy for Python inference? Any example?

**Reply by njzjz** [reply]
First you need to convert an existing model into the `.dp` format:
```sh
dp convert-backend aaa.pb bbb.dp
# or
dp convert-backend aaa.pth bbb.dp
```
Then `bbb.dp` can be used in `dp test`, `dp model-devi`, or any Python interface (`deepmd.infer.DeepPot`, `dpdata.System.predict`, etc).

**Reply by ShangChien** [reply]
To make a long story short, does this mean that we can get the energy and forces of the structure in the browser by pyodide-related projects 👀 ?
Previous related discussion https://github.com/deepmodeling/deepmd-kit/discussions/3100

**Reply by njzjz** [reply]
It cannot predict forces as numpy doesn't support automatic gradients, unless someone would like to write the backward functions manually.

**Reply by ShangChien** [reply]
Awesome DP!

**Comment by Epikurz** [comment]
Is there any examples for finetuning on DPA-2 pretrained model?

**Reply by ShangChien** [reply]
Several tutorial on [Bohrium website](https://nb.bohrium.dp.tech/cases), you can search with DPA tag to find them.
for example: [Notebook for DPA-2](https://nb.bohrium.dp.tech/detail/18475433825)
hope this helps 😸

**Reply by Epikurz** [reply]
Does It mean that all the functions of dp-pytorch version  have been inplemented in deepmd-kit 3?

**Reply by njzjz** [reply]
This repo was entirely imported in #3180.

**Comment by omidshy** [comment]
Nice work!

In the DPA-2 descriptor, the outputs of the `env` block are the rotational invariant `gij` and rotational equivariant `hij` representations. Does this make the DPA-2 an equivariant model?

**Comment by wanghan-iapcm** [comment]
> Nice work!
> 
> In the DPA-2 descriptor, the outputs of the `env` block are the rotational invariant `gij` and rotational equivariant `hij` representations. Does this make the DPA-2 an equivariant model?

When discussing the representation of a rotation group, specifically SO(3), it is necessary to introduce the concept of representation space. In the majority of "equivariant" models, this space is typically represented by spherical harmonics. To preserve as much information about the atomic environment as possible, one must increase the completeness (denoted as 'l' in the context of spherical harmonics) of the basis. However, this comes at the cost of increased computational expense.

In the development of DP models, the approach does not involve spanning the atomic environment within a representation space. Instead, it focuses on utilizing the original (equivariant) information without any approximation. This may be the reason why DPA-2 exhibits higher generalizability in our benchmarks, as evidenced by supplementary S2 of https://arxiv.org/abs/2312.15492.

**Comment by poyaii** [comment]
Hello. Thank you for developing the new version of DeepMd. When will the stable version of this release be available? Is it possible to install it via Conda? Thanks a lot.

**Reply by njzjz** [reply]
Conda packages are available at the `conda-forge/label/deepmd-kit_dev` channel: https://anaconda.org/conda-forge/deepmd-kit/files?version=3.0.0a0

There is no ETA for the stable version, but at least we need to resolve all issues in https://github.com/orgs/deepmodeling/projects/19
